### PR TITLE
support attn_dp_expert parallelism

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

  - Add attn_dp_expert mesh axis for expert parallelism: Introduce a new attn_dp_expert axis in the mesh and logical axis rules to support attention DP-aware expert parallelism, particularly when using 
  vllm_rpa attention.                                                                                                                                                                                     
  - Make expert axis name configurable in MoE layer: Replace hardcoded "expert" axis name in RoutedMoE with a configurable _expert_parallelism_name that switches to "attn_dp_expert" when attention ==   
  "vllm_rpa", affecting all collective operations (all_gather, psum_scatter, axis_index, ragged_all_to_all).                                                                                              
  - Add rollout_expert_parallelism config for RL training: New config field to specify expert parallelism per replica for rollout, with updated device count validation (tp * dp * ep ==
  num_sampler_devices).
  - Pass expert parallelism settings to vLLM rollout: When rollout_expert_parallelism > 1, pass expert_parallel_size and rollout_vllm_enable_expert_parallelism to vLLM kwargs.
  - Add vllm_config_path config option: New config field (in base.yml, rl.yml, and types.py) to specify the path to a YAML file for loading vLLM config, defaulting to
  src/maxtext/configs/inference/vllm.yml.
  - Update vLLM inference script: Add vllm_swap_space, vllm_async_scheduling, and vllm_config_path flags to vllm_decode.py; remove the hard requirement on hf_config_path.
  - Update vLLM logical axis rules: Add attn_dp_expert to sharding rules for activation axes (activation_q_length, activation_kv_batch, decode_batch, embed, exp, etc.) and add moe_mlp rule in the vLLM
  inference config.
  

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

DP=1,EP=64 Qwen3-235B-A22B on 2 slices of v5p-128: [command](https://docs.google.com/document/d/1kmcdLW9nbMIy2gH7z-dH1XE517t7MS4xl05OpGIOXx8/edit?tab=t.0#bookmark=kix.apic60a9rb1g)

Tested locally with Qwen3-30B and num_layers=5
```
NEW_MODEL_DESIGN=1 python src/maxtext/trainers/post_train/rl/train_rl.py src/maxtext/configs/post_train/rl.yml base_output_directory=gs://runner-maxtext-logs model_name=qwen3-30b-a3b override_model_config=true base_num_decoder_layers=5 tokenizer_path=Qwen/Qwen3-30B-A3B run_name=mohit-$RANDOM batch_size=4 num_batches=1 num_test_batches=1 rl.num_generations=2 max_target_length=260 max_prefill_predict_length=256 learning_rate=5e-7 ici_expert_parallelism=1 rollout_data_parallelism=1 rollout_tensor_parallelism=4 rollout_expert_parallelism=1 enable_dp_attention=True hbm_utilization_vllm=0.6 scan_layers=True async_scheduling=false allow_split_physical_axes=True vllm_additional_config='{maxtext_config: {model_name: qwen3-30b-a3b, override_model_config: true, base_num_decoder_layers: 5, allow_split_physical_axes: true, log_config: false, weight_dtype: bfloat16}}' vllm_hf_overrides='{architectures: ["MaxTextForCausalLM"], num_hidden_layers: 5}'
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -465,7 +470,12 @@ def _logical_to_mesh_axes(self, logical_name):
     return logical_to_mesh_axes(logical_name, mesh=self.mesh, rules=self.config.logical_axis_rules)
 
   def get_expert_parallelism_size(self):
-    return self.mesh.shape.get("expert", 1)
+    if isinstance(self._expert_parallelism_name, tuple):
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: Is the tuple case actually supported? Have you tested it?

Comment (khatwanimohit) [reply src/maxtext/layers/moe.py]: reverted this function, I was not assigning tuples to expert_parallelism_name. See [here](https://github.com/AI-Hypercomputer/maxtext/pull/3141/changes#diff-e080eae53990c3bcfd6cd1f60db63f57d80988dfe8715b4ffbc09fbcabb5d8f3R371-R375)

tuples don't work for functions like jax.lax.axis_index in sparse_matmul but they do work for other functions and we use tuples for `self._tensor_parallelism_name`

File: src/maxtext/trainers/post_train/rl/train_rl.py [hunk 2]
```diff
@@ -244,17 +245,22 @@ def get_rollout_kwargs_for_data_parallelism(sampler_config, num_sampler_devices)
...
     )
   rollout_kwargs["tensor_parallel_size"] = tp
   rollout_kwargs["data_parallel_size"] = dp
-  rollout_kwargs["rollout_vllm_async_scheduling"] = True
 
+  if ep > 1:
+    rollout_kwargs["expert_parallel_size"] = ep
+    rollout_kwargs["rollout_vllm_enable_expert_parallelism"] = True
+
+  rollout_kwargs["rollout_vllm_async_scheduling"] = sampler_config.rollout_async_scheduling
```
Comment (NicoGrande) [inline src/maxtext/trainers/post_train/rl/train_rl.py]: Can we remove this from this helper function and just assign it directly where we are defining the rollout config?

Comment (NicoGrande) [reply src/maxtext/trainers/post_train/rl/train_rl.py]: I mean `rollout_vllm_async_scheduling` specifically.

Comment (khatwanimohit) [reply src/maxtext/trainers/post_train/rl/train_rl.py]: removed rollout_vllm_async_scheduling

File: src/maxtext/trainers/post_train/rl/train_rl.py
Comment (xuefgu) [inline src/maxtext/trainers/post_train/rl/train_rl.py]: With the addition of EP, I think the new logic should be:

1. At most one of tp, dp and ep can be -1.
2. If one of them is -1, derive it by dividing `num_sampler_devices` by the other two (must be divisible).
3. If none of them is -1, the product has to be `num_sampler_devices`.

File: src/maxtext/trainers/post_train/rl/train_rl.py [hunk 3]
```diff
@@ -244,16 +245,20 @@ def get_rollout_kwargs_for_data_parallelism(sampler_config, num_sampler_devices)
...
         f"rollout_tensor_parallelism({tp}) * "
-        f"rollout_data_parallelism({dp}) "
+        f"rollout_data_parallelism({dp}) * "
+        f"rollout_expert_parallelism({ep}) "
         f"!= len(sampler_devices)({num_sampler_devices})"
     )
   rollout_kwargs["tensor_parallel_size"] = tp
   rollout_kwargs["data_parallel_size"] = dp
 
+  if ep > 1:
```
Comment (xuefgu) [inline src/maxtext/trainers/post_train/rl/train_rl.py]: Is `ep == 1` a special case that doesn't work? If not, this is not needed?

File: src/maxtext/trainers/post_train/rl/train_rl.py
Comment (khatwanimohit) [reply src/maxtext/trainers/post_train/rl/train_rl.py]: fixed the logic

File: src/maxtext/trainers/post_train/rl/train_rl.py (hunk 3 above)
Comment (khatwanimohit) [reply src/maxtext/trainers/post_train/rl/train_rl.py]: removed
